# Word Importance Classification

This code is main edited by Enze Wang

Fixed by Hua Tong, Yicen Liu

# Attention

Please put four .csv file in data folder and put this code in main folder instead of code folder.

This code want to provide you some important word list by tf-idf

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import random
import numpy as np
import nltk
import string
nltk.download('stopwords')
nltk.download('punkt')
from nltk.corpus import stopwords
from nltk.stem.wordnet import WordNetLemmatizer
lmtzr = WordNetLemmatizer()
import itertools
from collections import Counter
from nltk import FreqDist
from nltk.stem import PorterStemmer
nltk.download('wordnet')
from nltk.corpus import wordnet as wn
nltk.download('averaged_perceptron_tagger')
from wordcloud import WordCloud
from nltk import sent_tokenize, word_tokenize, PorterStemmer, pos_tag 
from nltk.corpus import stopwords
from nltk.text import TextCollection
from nltk.corpus import wordnet
import nltk.tokenize as nt
from sklearn.feature_extraction.text import CountVectorizer

[nltk_data] Downloading package stopwords to /Users/enze/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /Users/enze/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /Users/enze/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/enze/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


# Word Importance with TF-IDF

This code calculate word importance by TF-IDF, another code by Yicen is by tree regression.

TF-IDF includes two parts:

TF(WORD) = (Number of times term WORD appears in a document) / (Total number of terms in the document)

IDF(WORD) = ln(Total number of documents / Number of documents with term WORD in it)

TF-IDF=TF*IDF

Therefore high TF-IDF means that this word appears many times but not as many times as words like "and" and "is" because they are too common. 

TF-IDF only consider the word frequency but has no connection with star rating. 

For example, there are three documents:

"I have an apple"

"I have a pen"

"apple pen pen apple"


In [2]:
text_all=[["I","have","an","apple"],["I","have","a","pen"],["apple","pen","pen","apple"]]
text_collection = TextCollection(text_all)
for word in ["I","have","an","apple"]:
    print(word, "idf=", text_collection.idf(word))

I idf= 0.4054651081081644
have idf= 0.4054651081081644
an idf= 1.0986122886681098
apple idf= 0.4054651081081644


"I" occur in 2 documents, and there are 3 documents, therefore its idf is ln(3/2)=0.4054651

In [3]:
all_word=[w for text in text_all for w in text]
all_word

['I',
 'have',
 'an',
 'apple',
 'I',
 'have',
 'a',
 'pen',
 'apple',
 'pen',
 'pen',
 'apple']

In [4]:
for word in ["I","have","an","apple"]:
    word_tf = text_collection.tf(word, all_word)
    print(word, "tf=", word_tf)

I tf= 0.16666666666666666
have tf= 0.16666666666666666
an tf= 0.08333333333333333
apple tf= 0.25


"I" occur 2 times in 12 in the whole document, therefore its tf is 1/6=0.1666667

In [5]:
for word in ["I","have","an","apple"]:
    word_tf_idf = text_collection.tf_idf(word, all_word)
    print(word, "tf_idf=", word_tf_idf)

I tf_idf= 0.06757751801802739
have tf_idf= 0.06757751801802739
an tf_idf= 0.0915510240556758
apple tf_idf= 0.1013662770270411


So "I" tf-idf is 0.4054651081081644*0.16666666666666666=0.06757752

In [6]:
clean_review=pd.read_csv("data/pizza_business_clean_review.csv")

In [7]:
clean_review

Unnamed: 0               review_id  \
0               0  hfXVxrlD6fr1kXfz6sR13Q   
1               1  EDQpLLFiAxeSxWI5ipwzBQ   
2               2  RWmT90rP7wYo57dG1cyvww   
3               3  0jduztItUhOh3FzKGjMxZQ   
4               4  sxfmzhMvZ8LlI1x4fhNjKg   
...           ...                     ...   
67151       67151  gcNRaJMDlSVs8YlqwpsRuw   
67152       67152  yFxDOdiDb8HPaK1io81kiw   
67153       67153  Z_xmoLgIMB4As90t0a5eBQ   
67154       67154  CzYRFCnkbqOVb3WIMMrbfw   
67155       67155  ZhUeFHPAjUbEKRnYO_pl_w   

                                            clean_review  
0      parking tough get dont get go much also get li...  
1      server extensive knowledge wine various menu s...  
2      fast forward every trip nyc pitt includes visi...  
3      far favorite clevelandakron restaurant great f...  
4      food good usually get italian sub pizza would ...  
...                                                  ...  
67151  im originally side ohio ive always really enjo...  
67152  yes yes yes loved going place shadyside ecstat...  
67153  went lunch meeting amidst typical suburban str...  
67154  wife eaten la brioche least dozen time food ne...  
67155  first let say lindy columbus one favorite rest...  

[67156 rows x 3 columns]

In [8]:
text_all=[]
r,c=clean_review.shape
for i in range(r):
    text_all.append(list(map(str,clean_review["clean_review"][i].split(" "))))
    

In [16]:
text_all

[['parking',
  'tough',
  'get',
  'dont',
  'get',
  'go',
  'much',
  'also',
  'get',
  'little',
  'cautious',
  'sometimes',
  'frequently',
  'multiple',
  'questionable',
  'looking',
  'character',
  'pro',
  'legendary',
  'fry',
  'used',
  'super',
  'huge',
  'gave',
  'great',
  'value',
  'buck',
  'good',
  'burger',
  'drink',
  'look',
  'like',
  'lot',
  'different',
  'kind',
  'alcohol',
  'con',
  'customer',
  'service',
  'hahaha',
  'get',
  'fart',
  'one',
  'twin',
  'working',
  'ahead',
  'pack',
  'dirty',
  'messy',
  'table',
  'crap',
  'everywhere',
  'many',
  'questionable',
  'character',
  'someone',
  'asked',
  'wanted',
  'buy',
  'stolen',
  'jewelry',
  'eating',
  'im',
  'totally',
  'honest',
  'french',
  'fry',
  'portion',
  'gotten',
  'smaller',
  'used',
  'need',
  'two',
  'basket',
  'large',
  'fry',
  'need',
  'one',
  'really',
  'expensive',
  'hot',
  'dog',
  'strict',
  'limit',
  'ketchup',
  'im',
  'joking',
  'large',


In [9]:
all_word=[w for text in text_all for w in text]

In [10]:
all_word

['parking',
 'tough',
 'get',
 'dont',
 'get',
 'go',
 'much',
 'also',
 'get',
 'little',
 'cautious',
 'sometimes',
 'frequently',
 'multiple',
 'questionable',
 'looking',
 'character',
 'pro',
 'legendary',
 'fry',
 'used',
 'super',
 'huge',
 'gave',
 'great',
 'value',
 'buck',
 'good',
 'burger',
 'drink',
 'look',
 'like',
 'lot',
 'different',
 'kind',
 'alcohol',
 'con',
 'customer',
 'service',
 'hahaha',
 'get',
 'fart',
 'one',
 'twin',
 'working',
 'ahead',
 'pack',
 'dirty',
 'messy',
 'table',
 'crap',
 'everywhere',
 'many',
 'questionable',
 'character',
 'someone',
 'asked',
 'wanted',
 'buy',
 'stolen',
 'jewelry',
 'eating',
 'im',
 'totally',
 'honest',
 'french',
 'fry',
 'portion',
 'gotten',
 'smaller',
 'used',
 'need',
 'two',
 'basket',
 'large',
 'fry',
 'need',
 'one',
 'really',
 'expensive',
 'hot',
 'dog',
 'strict',
 'limit',
 'ketchup',
 'im',
 'joking',
 'large',
 'fry',
 'anything',
 'le',
 'recommend',
 'getting',
 'legendary',
 'fry',
 'least',
 '

In [11]:
important_word=pd.read_csv("data/Word_Freq.csv")

In [12]:
important_word

Unnamed: 0         Word  Frequency
0              0        pizza      97915
1              1         good      40169
2              2        place      36894
3              3         food      34945
4              4        great      29749
...          ...          ...        ...
5490        5490  awesomeness         31
5491        5491    scrambled         31
5492        5492     woodfire         31
5493        5493         coin         31
5494        5494        hired         31

[5495 rows x 3 columns]

# Attention:  Code running for long time below:

In [13]:
Word_list=important_word["Word"].tolist()
text_collection = TextCollection(text_all)
TF_IDF=[]
for word in Word_list:
    word_tf_idf = text_collection.tf_idf(word, all_word)
    TF_IDF.append(word_tf_idf)
important_word["tf_idf"]=TF_IDF

In [14]:
important_word=important_word.sort_values(by=['tf_idf'],ascending=False)
important_word[0:20]

Unnamed: 0     Word  Frequency    tf_idf
0            0    pizza      97915  0.014674
1            1     good      40169  0.011126
3            3     food      34945  0.010568
2            2    place      36894  0.010550
5            5     time      27346  0.010237
7            7    order      21334  0.010055
4            4    great      29749  0.010030
6            6     like      22193  0.009456
8            8      one      20664  0.009227
10          10      get      19846  0.008941
11          11    would      17730  0.008596
16          16   cheese      16161  0.008584
14          14    sauce      16568  0.008568
12          12  ordered      17604  0.008510
17          17   really      15991  0.008221
9            9  service      20018  0.008180
13          13     back      17495  0.008032
20          20    salad      13152  0.007915
15          15       go      16229  0.007828
18          18     also      14240  0.007517

As you can see, there is totally no difference between the rank with frequence and the rank with TF-IDF. The reason is that the stop words has been removed from the steps above.


In [15]:
importance_word=important_word.drop(["Unnamed: 0"],axis=1)
importance_word.to_csv("data/Importance_word_tf_idf.csv")